### ISSUE TITLE
[SUGGESTION] Create docker images for participants to use

### ISSUE THREAD

**Issue by SyedAman** [exercism/elixir#397]
We want to improve the user experience for all Exercism users. We want to make it as easy as possible to get up and running solving exercises without the frustration. Windows users usually run into some sort of installation problems too.

One way we can achieve this is by setting up environments ourselves for them to use via Docker! So in the Elixir track, we can provide participants a Docker image to use. Sometimes users also don't want to install too many packages in their machine to avoid cluttering.

Users would just need to type in something like:

```bash
$ docker run -it -v $(pwd):/root/exercism/elixir exercism-elixir
```

And get started asap!

The Dockerfile can be defined as such in an official registry on DockerHub for Exercism:

```docker
FROM elixir

# Download and install Exercism CLI
RUN wget https://github.com/exercism/cli/releases/download/v2.4.1/exercism-linux-64bit.tgz
RUN tar -xzvf exercism-linux-64bit.tgz
RUN mkdir ~/bin
RUN mv exercism ~/bin/
RUN export PATH=$HOME/bin:$PATH
RUN echo 'export PATH=$HOME/bin:$PATH' >> ~/.bashrc

# Create working directory
WORKDIR /root/exercism/elixir
VOLUME /root/exercism/elixir

ENTRYPOINT ~/bin/exercism fetch elixir && bash
```

**Comment by NobbZ** [comment]
1. I'd not install `exercism` into `${HOME}/bin` but into `/usr/local/bin`.
2. The resulting image is bloated due to the nature of overlayed filesystems. Do the following instead of the single command `RUN`s you have.
  ```dockerfile
  RUN set -ex; \
      oldwd=$(pwd)
      mkdir -p /tmp/exercism-install; \
      cd /tmp/exercism-install; \
      wget https://github.com/exercism/cli/releases/download/v2.4.1/exercism-linux-64bit.tgz; \
      tar -xzvf exercism-linux-64bit.tgz; \
      mv exercism /usr/local/bin/exercism; \
      cd $oldwd; \
      rm -rfv /tmp/exercism-install
  ```
3. Provide `ARG` fields to create exercism configuration on build.
4. AFAIK you can not use `&&` in an `ENTRYPOINT`
5. I'd create a proper user inside the container
6. I'd pin the elixir version you derive the image `FROM`
7. I'd go for an `*-alpine` base image to reduce download size. This would require you to build `exercism` CLI from source though, since the version supplied from the releases page is dynamically linked to `glibc` while alpine uses `libmusl`.
8. Provide documentation about how this image is expected to be used.

In summary, I do not think your current Dockerfile is mature enough to propose it publicly.

**Comment by SyedAman** [comment]
@NobbZ This is great feedback. This Dockerfile is not mature enough, but it can be with help from people like you.

> I'd not install exercism into ${HOME}/bin but into /usr/local/bin.
The resulting image is bloated due to the nature of overlayed filesystems.

If this is indeed true, then the install instructions need to be updated: http://exercism.io/clients/cli/linux. 

> Do the following instead of the single command RUNs you have.
RUN set -ex; \
    oldwd=$(pwd)
    mkdir -p /tmp/exercism-install; \
    cd /tmp/exercism-install; \
    wget https://github.com/exercism/cli/releases/download/v2.4.1/exercism-linux-64bit.tgz; \
    tar -xzvf exercism-linux-64bit.tgz; \
    mv exercism /usr/local/bin/exercism; \
    cd $oldwd; \
    rm -rfv /tmp/exercism-install

This is good for a clean install of exercism, nice.

> RUN set -ex; \

What does `set -ex` do? I can't seem to google it.

> AFAIK you can not use && in an ENTRYPOINT

You most certainly can, perhaps it's just poor style. But `ENTRYPOINT` is not neccessary, using `CMD` instead.

> I'd pin the elixir version you derive the image FROM

Elixir version `1.6.4` is now pinned.

> I'd go for an *-alpine base image to reduce download size. This would require you to build exercism CLI from source though, since the version supplied from the releases page is dynamically linked to glibc while alpine uses libmusl.

Good idea. The base image is now an elixir base image using alpine. This will help reduce download size indeed. It seems to still work without building exercism CLI from source?

> Provide ARG fields to create exercism configuration on build.

Yeah this is a great idea. I was thinking about this earlier. I need to figure out how to do this now...

> I'd create a proper user inside the container.

The user probably wants to pass in the user name in `ARG`?

**Comment by SyedAman** [comment]
I should perhaps make a PR for this

**Comment by NobbZ** [comment]
> If this is indeed true, then the install instructions need to be updated: http://exercism.io/clients/cli/linux.

The installation instructions do assume a natively running linux, not docker.

For Docker some special casing applies that one needs to be aware of when creating dockerfiles.

In docker each `RUN` creates a layer, which is basically a delta of the previous filesystem. If you now download and extract in two separate `RUN` statements, then you'll have those layers immutably with that data.

Then you move around the binary, creating a layer telling “delte A, create B”, so this layer is as big as the previous layer, and even when deleting the tarball in just another run, this will not remove the layer that initially created the tarball, and each layer that has been created needs to be stored on disc or even worse, it needs to get downloaded if not building the image on your own computer.

Please read [docker best practices](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/) its explained there as well as a lot of other stuff.

> What does `set -ex` do? I can't seem to google it.

It sets some options in the shell. In combination they cause the script to fail instantly as soon as a step inbetween fails, also it causes to print each command when run. I can't remember though which of the option is causing which effect ;) I tend to use them as they are everywhere ;)

> You most certainly can, perhaps it's just poor style. But ENTRYPOINT is not neccessary, using CMD instead.

I'd not `exercism f elixir` at all. I'd prefer if it were more explicit.

----

Also I'd take a look into the completion scripts for bash and zsh and put them in place while using bash or zsh as entrypoint.

----

PS: If done correctly, there is no need to alter `$PATH`, especially `/usr/local/bin` should already be in there.

This path reserved for system wide binaries that are not managed by the distributions base repositories. In most configurations it takes precedence over `/usr/bin` which is for files managed by the distros base repositories and `/bin` which is meant for stuff even closer to the core… (At least this is the hirarchy as I understood it when when we had our linux workshop at university 3 years ago).

**Comment by devonestes** [comment]
Just so I know, do you intend this to be helpful for people to more easily contribute to this repository, or to more easily do the exercises?

The exercises run without any dependencies other than Erlang and Elixir available on the system being used, and the only dependency we have for development is dialyxer, and that's just a hook into something that's built into Erlang. So I'm not really worried about clutter here.

Is there some specific pain point that you had while trying to get started you can share?

**Comment by NobbZ** [comment]
Solving exercises probably:

> We want to improve the user experience for all Exercism users. We want to make it as easy as possible to get up and running solving exercises without the frustration. Windows users usually run into some sort of installation problems too.

**Comment by NobbZ** [comment]
PS: The only hassle I had when I installed elixir on a windows machine was setting up `PATH` correctly, but that has been somewhere around 1.0 or 1.1.

Also for the elixir executable one has to set up `PATH`

Thats a one time thing for each of those.

For docker one needs to set up the PATH, the underlying VM (on windows at least), and probably the shared volumes. Also one needs to remember and retype the `docker run` command with the pathes to the volume etc. with the big downsite, that there is no replacement for shells `$()` on windows, so the command from the OP won't even work on WIN (and thats where installation of elixir is most complicated, but less complicated than docker).

As I said in #398 already, I do not think such an image would help anyone.

Maybe have it as an additional service for folks who already know how to work with docker, but not as "here this is easier for everyone".

**Comment by devonestes** [comment]
Yeah, I think no matter what, you have to install some software to make this work. If folks are experienced enough to install Docker on their windows machine, I think they're more than experienced enough to install elixir and erlang.

I'm gonna go ahead and close this since I don't think this is a great way to roll.
